sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1450,7 +1450,9 @@ object CleanupAliases extends Rule[LogicalPlan] {
         case c: CreateStructUnsafe if !stop =>
           stop = true
           c.copy(children = c.children.map(trimNonTopLevelAliases))
-        case Alias(child, _) if !stop => child
+        // Only eliminate aliases for named expressions, otherwise we may turn an `Alias` to a
+        // normal expression and break the type requirement for it.
+        case Alias(child: NamedExpression, _) if !stop => child
```

Comment:
Should we add the other operators that can produce attributes above instead of changing this?  I'm not sure.